### PR TITLE
Fix hack/update-deps.sh to ignore go_binary rules in vendor

### PR DESCRIPTION
**Pull request by fejta**
Ask bazel to find all go libraries that are have no indirect dependency by anything outside vendor

* Aka `bazel query 'kind("go_library rule", //vendor/... -deps(//... -//vendor/...))'`
* This makes prune-libraries.sh run in around a minute instead of 20+

We are not dropping the unused `//vendor/github.com/golang/protobuf/protoc-gen-go` because of the zombie main package in this folder.

Fix this by ignoring `go_binary` rdeps in `//vendor`

Ran `hack/update-deps.sh && bazel test //...` with this change

/assign @stevekuznetsov @BenTheElder 
/cc @kargakis

### CODE REVIEW

File: hack/update-deps.sh [hunk 1]
```diff
@@ -55,6 +55,8 @@ drop-dep vendor/golang.org/x/text/language vendor/golang.org/x/text/internal
 drop-dep vendor/google.golang.org/api/transport/grpc vendor/google.golang.org/api/transport
 drop-dep vendor/github.com/golang/protobuf/protoc-gen-go/grpc vendor/github.com/golang/protobuf/protoc-gen-go
 drop-dep vendor/github.com/golang/protobuf/protoc-gen-go/generator vendor/github.com/golang/protobuf/protoc-gen-go
+drop-dep vendor/github.com/golang/protobuf/protoc-gen-go/generator vendor/github.com/golang/protobuf/protoc-gen-go
```
Comment (stevekuznetsov) [inline hack/update-deps.sh]: This line is a dupe of the one above it?

Comment (fejta) [reply hack/update-deps.sh]: Done

**Comment by BenTheElder** [comment]
`hack/update-deps.sh` also only considers bazel so it drops `//vendor/github.com/gogo/protobuf/proto` since there's just a compiled `foo.pb.go` and a proto importing this library in `//vendor/github.com/docker/docker/api/types/swarm/runtime` (which gazelle inserts it's own `go_proto_library` rule to build). The docker API (and therefore barnacle) can only be bazel built now.

```
$ go build ./images/bootstrap/barnacle
vendor/github.com/docker/docker/api/types/swarm/runtime/plugin.pb.go:17:8: cannot find package "github.com/gogo/protobuf/proto" in any of:
        /usr/local/google/home/bentheelder/go/src/k8s.io/test-infra/vendor/github.com/gogo/protobuf/proto (vendor tree)
        /usr/lib/google-golang/src/github.com/gogo/protobuf/proto (from $GOROOT)
        /usr/local/google/home/bentheelder/go/src/github.com/gogo/protobuf/proto (from $GOPATH)
```
 versus
```
$ bazel build //images/bootstrap/barnacle:barnacle
..........
INFO: Analysed target //images/bootstrap/barnacle:barnacle (213 packages loaded).
INFO: Found 1 target...
Target //images/bootstrap/barnacle:barnacle up-to-date:
  bazel-bin/images/bootstrap/barnacle/linux_amd64_stripped/barnacle
INFO: Elapsed time: 12.964s, Critical Path: 3.02s
INFO: Build completed successfully, 1 total action
```
:man_shrugging:

**Comment by ixdy** [comment]
I think gazelle is going to stop generating proto rules for `vendor/`, though, which might break this.

**Comment by ixdy** [comment]
Indeed, if I take test-infra, update `gazelle` to https://github.com/bazelbuild/bazel-gazelle/commit/2f186389e2d9a91ee64007914f9b9d0ecae1d8e9, and run `hack/update-bazel.sh`, then

```diff
gazelle: multiple rules (//vendor/github.com/docker/docker/api/types/swarm/runtime:runtime_go_proto and //vendor/github.com/docker/docker/api/types/swarm/runtime:go_default_library) may be imported with "github.com/docker/docker/api/types/swarm/runtime" from //vendor/github.com/docker/docker/api/types/swarm:go_default_library
--- [elided]/src/k8s.io/test-infra/vendor/github.com/docker/docker/api/types/swarm/runtime/BUILD        2018-01-23 12:07:22.289526798 -0800
+++ /tmp/BUILD603063607 2018-01-23 12:08:27.349735222 -0800
@@ -16,10 +16,13 @@

 go_library(
     name = "go_default_library",
-    srcs = ["gen.go"],
-    embed = [":runtime_go_proto"],
+    srcs = [
+        "gen.go",
+        "plugin.pb.go",
+    ],
     importpath = "github.com/docker/docker/api/types/swarm/runtime",
     visibility = ["//visibility:public"],
+    deps = ["//vendor/github.com/gogo/protobuf/proto:go_default_library"],
 )
 
 filegroup(
--- [elided]/src/k8s.io/test-infra/vendor/github.com/docker/docker/api/types/swarm/BUILD        2018-01-23 12:07:22.285526786 -0800
+++ /tmp/BUILD671893290 2018-01-23 12:08:27.349735222 -0800
@@ -20,7 +20,6 @@
         "//vendor/github.com/docker/docker/api/types/container:go_default_library",
         "//vendor/github.com/docker/docker/api/types/mount:go_default_library",
         "//vendor/github.com/docker/docker/api/types/network:go_default_library",
-        "//vendor/github.com/docker/docker/api/types/swarm/runtime:go_default_library",
     ],
 )
```

**Comment by kargakis** [comment]
@ixdy I am updating gazelle in https://github.com/kubernetes/test-infra/pull/6199 exactly for dropping proto generation from BUILD files. Mind chiming in as to whether we can bump gazelle?

**Comment by ixdy** [comment]
I thought I remembered seeing that somewhere. I'm fine with bumping gazelle, though I have no idea what it'll do to this PR.

**Comment by BenTheElder** [comment]
Maybe we should start with a smaller PR to bump gazelle?

**Comment by fejta** [comment]
I don't see any reason not to get this in

File: hack/update-deps.sh (hunk 1 above)
Comment (ixdy) [reply hack/update-deps.sh]: did you forget to push? I still see the duplicate here.

**Comment by fejta** [comment]
PTAL

**Comment by stevekuznetsov** [comment]
Dupe line still there? Otherwise

File: hack/update-deps.sh (hunk 1 above)
Comment (fejta) [reply hack/update-deps.sh]: Done. So confused...

**Comment by fejta** [comment]
Sorry missed Jeff's comment. updated!

File: hack/prune-libraries.sh [hunk 2]
```diff
@@ -25,30 +25,8 @@ set -o errexit
 set -o pipefail
 
 # Output every go_library rule in the repo
```
Comment (ixdy) [inline hack/prune-libraries.sh]: this comment is outdated :)

Comment (fejta) [reply hack/prune-libraries.sh]: Done

**Comment by stevekuznetsov** [comment]
/lgtm
🚢 🤡 
:shipit:

**Pull request merged**